pennylane_lightning/lightning_qubit/lightning_interpreter.py

```diff
@@ -0,0 +1,170 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains a class for executing plxpr using default qubit tools.
+"""
+from copy import copy
+from functools import partial
+from typing import Optional
+
+import jax
+from pennylane.capture import disable, enable
+from pennylane.capture.base_interpreter import PlxprInterpreter
+from pennylane.capture.primitives import (
+    adjoint_transform_prim,
+    cond_prim,
+    ctrl_transform_prim,
+    for_loop_prim,
+    measure_prim,
+    while_loop_prim,
+)
+from pennylane.measurements import MidMeasureMP, Shots
+
+from ._measurements import LightningMeasurements
+from ._state_vector import LightningStateVector
+
+
+class LightningInterpreter(PlxprInterpreter):
+    """
+
+
+
+    .. code-block:: python
+
+        import pennylane as qml
+        import jax
+        qml.capture.enable()
+
+        statevector = LightningStateVector(num_wires=3, dtype=np.complex128)
+
+        @LightningInterpreter(statevector, shots=Shots(None))
+        def f(x):
+            @qml.for_loop(3)
+            def loop(i, y):
+                qml.RX(y, i)
+                return y + 0.5
+            loop(x)
+            return [qml.expval(qml.Z(i)) for i in range(3)]
+
+        f(0.0)
+
+    .. code-block::
+
+        [1.0, 0.8775825618903728, 0.5403023058681395]
+    """
+
+    def __init__(self, state: LightningStateVector, shots: Shots = Shots(None)):
+        self.state = state
+        self.shots = shots
+        if self.shots.has_partitioned_shots:
+            raise NotImplementedError("LightningInterpreter does not support partitioned shots.")
+        self.reset = True
+        super().__init__()
+
+    def setup(self) -> None:
+        """Reset the state if necessary."""
+        if self.reset:
+            self.state.reset_state()
+            self.reset = False  # copies will have reset=False and wont reset state
+
+    def cleanup(self) -> None:
+        """Indicate that the state will need to be reset if this instance is reused."""
+        self.reset = True
+
+    def interpret_operation(self, op):
+        """Apply an operation to the state."""
+        self.state.apply_operations([op])
+
+    def interpret_measurement_eqn(self, eqn: "jax.core.JaxprEqn"):
+        """Interpret a given measurement equation."""
+        if "mcm" in eqn.primitive.name:
+            raise NotImplementedError(
+                "DefaultQubitInterpreter does not yet support postprocessing mcms"
+            )
+        return super().interpret_measurement_eqn(eqn)
+
+    def interpret_measurement(self, measurement):
+        """Apply a measurement to the state and return numerical results."""
+        # measurements can sometimes create intermediary mps, but those intermediaries will not work with capture enabled
+        disable()
```

Comment:
I don't want to promote potentially bad coding patterns, but this as a context manager could be a "hacky" way to create a non-traceable context. Though maybe that's just a `pure_callback` 😅

Reply:
I think we need a way using `process_state` and `process_samples` that doesn't use intermediary measurements. Or a way of capture-safe producing intermediary measurements.

Reply:
Unrelated to this PR, but maybe we can create a separate function that performs the operations of `some_measurement._primitive.impl`, and we can call that function inside `primitive.impl`, and also here.